Refine VAA page (#293)
wormhole-foundation/wormhole-docs: learn/infrastructure/vaas.md

# Verified Action Approvals

Verified Action Approvals (VAAs) are Wormhole's core messaging primitive. They are packets of cross-chain data emitted whenever a cross-chain application contract interacts with the Core Contract.

[Guardians](/docs/learn/infrastructure/guardians/){target=\_blank} validate messages emitted by contracts before sending them to the target chain. Once a majority of Guardians agree the message is valid, they sign a keccak256 hash of the message body. 

The message is wrapped up in a structure called a VAA, which combines the message with the Guardian signatures to form a proof. 

VAAs are uniquely indexed by the (`emitter_chain`, `emitter_address`, `sequence`) tuple. To obtain a VAA, one can query the [Wormholescan API](https://docs.wormholescan.io/){target=\_blank} with this information.

> !!!tip "Sequence number"

Reviewer: This feels like contextual information rather than helpful advice. I'm not sure. What's our criteria for deciding if it should be a note or a tip?

    The `sequence` field depends on the final ordering of blocks on the emitter chain. When a lower consistency level is chosen (i.e., not waiting for finality), there is a chance that chain reorganizations could lead to multiple, different VAAs appearing for what looks like the “same” message on the user side. 
    
    The tuple (`emitter_chain`, `emitter_address`, `sequence`) can only be considered unique if the chain does not undergo a reorg and the block containing the message has effectively reached finality. However, there is always a small chance of an extended reorg that could invalidate or alter a previously emitted sequence number.

## VAA Format

The basic VAA consists of header and body components described as follows:

> - **Header** - holds metadata about the current VAA, the Guardian set that is currently active, and the list of signatures gathered so far

Reviewer: Nice

    - `version` ++"byte"++ - the VAA Version
    - `guardian_set_index` ++"u32"++ - indicates which Guardian set is signing
    - `len_signatures` ++"u8"++ - the number of signatures stored
    - `signatures` ++"[]signature"++ - the collection of Guardian signatures

    Where each `signature` is:

    - `index` ++"u8"++ - the index of this Guardian in the Guardian set
    - `signature` ++"[65]byte"++ - the ECDSA signature

- **Body** - _deterministically_ derived from an on-chain message. Any two Guardians processing the same message must derive the same resulting body to maintain a one-to-one relationship between VAAs and messages to avoid double-processing messages

!!!tip "Consistency level"

>     This is only strictly true once the chain's state is finalized. If a reorg occurs, and a transaction that previously appeared in block X is replaced by block Y, Guardians observing different forks may generate different VAAs for what the emitter contract believes is the same message.

Reviewer: What is only strictly true?

Reviewer 2: this should be about the body not the consistency level, i remember someone from wormhole asked me to write this 

the body is deterministically  derived from the on-chain message. Any two Guardians processing the same message must derive the same resulting body - this is only strictly true once the chain's state is finalized, If a reorg occurs Guardians observing different forks may generate different VAAs

Reviewer 2: the title of this note should be either note or important, not consistency level or any other field

    
The tuple (`emitter_chain`, `emitter_address`, `sequence`) can only be considered unique if the chain does not undergo a reorg and the block containing the message has effectively reached finality. However, there is always a small chance of an extended reorg that could invalidate or alter a previously emitted sequence number.

## VAA Format

The basic VAA consists of header and body components described as follows:

- **Header** - holds metadata about the current VAA, the Guardian set that is currently active, and the list of signatures gathered so far
    - `version` ++"byte"++ - the VAA Version
    - `guardian_set_index` ++"u32"++ - indicates which Guardian set is signing
    - `len_signatures` ++"u8"++ - the number of signatures stored
    - `signatures` ++"[]signature"++ - the collection of Guardian signatures

    Where each `signature` is:

    - `index` ++"u8"++ - the index of this Guardian in the Guardian set
    - `signature` ++"[65]byte"++ - the ECDSA signature

> - **Body** - _deterministically_ derived from an on-chain message. Any two Guardians processing the same message must derive the same resulting body to maintain a one-to-one relationship between VAAs and messages to avoid double-processing messages. 

Reviewer: ```suggestion
- **Body** - _deterministically_ derived from an on-chain message. Any two Guardians processing the same message must derive the same resulting body to maintain a one-to-one relationship between VAAs and messages to avoid double-processing messages
```

    - The deterministic nature of the body is only strictly true once the chain's state is finalized
    - If a reorg occurs, and a transaction that previously appeared in block X is replaced by block Y, Guardians observing different forks may generate different VAAs for what the emitter contract believes is the same message 

>     - This scenario is less likely once a block is sufficiently buried, but it can still happen if you choose a faster (less finalized) consistency level

Reviewer 2: i get that breaking this down might be more digestible but we have lists above and under, I'm not sure I like it, I'd prefer like a paragraph or a note since its an important thing to not overlook

Author: I made this more like the header one per Erin's suggestion and made those explaining bullets into a paragraph. Let me know what you think!

Reviewer 2: looks good in a paragraph!

    Elements of the body are as follows:

>     - `timestamp` ++"u32"++ - the timestamp of the block this message was published in

Reviewer: I saw Ilaria's comment so I came back to revisit this page. I do like how we have the header set up with a brief definition and then the elements that make up the header below it.

We should at least reorganize this to match so these elements are right after the definition of the body

### Token Transfer

Many bridges use a lockup/mint and burn/unlock mechanism to transfer tokens between chains. Wormhole's generic message-passing protocol handles the routing of lock and burn events across chains to ensure Wormhole's Token Bridge is chain-agnostic and can be rapidly integrated into any network with a Wormhole contract.

Transferring tokens from Chain A to Chain B requires the following steps:

- Lock the token on Chain A
- Chain A emits a message as proof the token lockup is complete
- Chain B receives the message confirming the lockup event on Chain A

> - Token is minted on Chain B

Reviewer: Nice! Use steps for sequential lists

- `symbol` ++"[32]byte"++ - short name of asset
- `name` ++"[32]byte"++ - full name of asset

#### Attestation Tips 

Be aware of the following considerations when working with attestations:

- Attestations use a fixed-length byte array to encode UTF8 token name and symbol data. Because the byte array is fixed length, the data contained may truncate multibyte Unicode characters

> - When sending an attestation VAA, it is recommended to send the longest UTF-8 prefix that doesn't truncate a character and then right-pad it with zero bytes

Reviewer: ```suggestion
- When sending an attestation VAA, it is recommended to send the longest UTF8 prefix that doesn't truncate a character and then right-pad it with zero bytes
```

The other instances on this page aren't using a hyphen

- When parsing an attestation VAA, it is recommended to trim all trailing zero bytes and convert the remainder to UTF-8 via any lossy algorithm

- Be mindful that different on-chain systems may have different VAA parsers, resulting in different names/symbols on different chains if the string is long or contains invalid UTF8

> - Without knowing a token's decimal precision, Chain B cannot correctly mint the number of tokens when processing a transfer. For this reason, the Token Bridge requires an attestation for each token transfer

Reviewer: I feel like we should say the destination chain instead of chain B in this case

### Token Transfer with Message

The Token Transfer with Message data structure is identical to the token-only data structure, except for the following:

- **`fee` field** - replaced with the `from_address` field 
- **`payload` field** - is added containing arbitrary bytes. A dApp may include additional data in this arbitrary byte field to inform some application-specific behavior

> This VAA type is also referred to as a `payload3` message or a Contract Controlled Transfer and has the following structure:

Reviewer: We have a bigger problem with Contract-Controlled Transfer more commonly being used instead of "Token Transfer with Message" - we should discuss this and update here accordingly because I think there are a few links to this section with the link text being "Contract Controlled Transfers" cc: @ilariae

Reviewer 2: i think contract controlled transfer came from their old docs but I don't remember exactly where from, I like token transfer with payload/message better it's more intuitive

Reviewer: Yeah, I agree with you

Reviewer: Made quick PR to update other instances: https://github.com/wormhole-foundation/wormhole-docs/pull/298

With the concepts now defined, it is possible to illustrate a full flow for message passing between two chains. The following stages demonstrate each step of processing that the Wormhole network performs to route a message.

1. **A message is emitted by a contract running on Chain A** - any contract can emit messages, and the Guardians are programmed to observe all chains for these events. Here, the Guardians are represented as a single entity to simplify the graphics, but the observation of the message must be performed individually by each of the 19 Guardians
2. **Signatures are aggregated** - Guardians independently observe and sign the message. Once enough Guardians have signed the message, the collection of signatures is combined with the message and metadata to produce a VAA
3. **VAA submitted to target chain** - the VAA acts as proof that the Guardians have collectively attested the existence of the message payload. The VAA is submitted (or relayed) to the target chain to be processed by a receiving contract and complete the final step

> ![Lifetime of a message diagram](/docs/images/learn/infrastructure/vaas/lifetime-vaa-diagram.webp)

Reviewer 2: the diagram looks really small it's not readable without opening it, maybe take it out from the list?

Reviewer 2: @dawnkelly09 i didn't mean to remove the diagram completely, my bad, just to not have it nested in the list so it gets a bit larger and maybe becomes more readable

Author: Sorry! I will fix!

Reviewer 2: i see, for now let's leave it like this, I'll follow up and maybe redo the diagram or ask their design team to make it more readable

## Lifetime of a Message

> Anyone can submit the VAA to the target chain. The Guardians typically don't perform this step to avoid transaction fees. Instead, applications built on top of Wormhole can acquire the VAA via the Guardian RPC and submit it in a separate flow.

Reviewer: ```suggestion
Anyone can submit a VAA to the target chain. Guardians typically don't perform this step to avoid transaction fees. Instead, applications built on top of Wormhole can acquire a VAA via the Guardian RPC and submit it in a separate flow.
```
